lib/http2/connection.c

```diff
@@ -804,7 +804,7 @@ static int handle_settings_frame(h2o_http2_conn_t *conn, h2o_http2_frame_t *fram
         }
         /* apply the change to window size (to all the streams but not the connection, see 6.9.2 of draft-15) */
         if (prev_initial_window_size != conn->peer_settings.initial_window_size) {
-            ssize_t delta = conn->peer_settings.initial_window_size - prev_initial_window_size;
+            ssize_t delta = (ssize_t) conn->peer_settings.initial_window_size - prev_initial_window_size;
```

Comment:
Thank you for reporting the issue.

Does https://github.com/h2o/h2o/commit/aebdb581666b51da95e538ded49175ff830729cd look like a fix to you? I think we need to cast both operands. Otherwise, if `ssize_t` is a 32-bit number, the subtraction would be carried in `uint32_t` and the result would continue to be erroneous.

Reply:
> if ssize_t is a 32-bit number, the subtraction would be carried in uint32_t

Thanks, that’s right. I assumed the type ssize_t is greater than 32-bit.
The case 32-bit `sszie_t` arithmetic conversions is well-described in the following docs 
https://wiki.sei.cmu.edu/confluence/display/c/INT02-C.+Understand+integer+conversion+rules

>> If the operand that has unsigned integer type has rank greater than or equal to the rank of the type of the other operand, the operand with signed integer type is converted to the type of the operand with unsigned integer type.

----

> the result would continue to be erroneous.

Again, right. Usually safe but it’s `implementation-defined`
ref: https://wiki.sei.cmu.edu/confluence/display/c/INT16-C.+Do+not+make+assumptions+about+representation+of+signed+integers

----

> Does aebdb58 look like a fix to you?

Yes, it does fix.
As you mentioned in the comment, `window sizes are guaranteed to be in 0..INT32_MAX`,  how about changing `initial_window_size` type from `uint32_t` into `int32_t` like I mentioned https://github.com/h2o/h2o/pull/1526#issuecomment-350160898 ? Casting in the subtraction seems to be a little complicated to me.

Reply:
> how about changing `initial_window_size` type from `uint32_t` into `int32_t` like I mentioned https://github.com/h2o/h2o/pull/1526#issuecomment-350160898?

Generally speaking, C does not provide a type that can store result of any integer arithmetic. We need to be sure that each calculation we code do not cause unexpected issues due to overflow.

In other words, changing the definition of `initial_window_size` to `int32_t` does not solve the issue. Therefore, I'd prefer being explicit; i.e. use explicit cast when we calculate.

Note also that RFC 7540 defines each settings variable as unsigned 32-bit variables.

Reply:
> RFC 7540 defining each settings variable as unsigned 32-bit variables.
> ...
> Considering the aspects, I'd prefer being natural and explicit;

According to https://tools.ietf.org/html/rfc7540#section-6.5.1 , it sounds natural to me that `uint32_t` definition and using explicit cast to both operands.